ProcessLib/TwoPhaseFlowWithPrho/TwoPhaseFlowWithPrhoMaterialProperties.h

```diff
@@ -0,0 +1,278 @@
+        std::unique_ptr<MaterialLib::Fluid::FluidProperty>
+            gas_viscosity,
+        std::vector<Eigen::MatrixXd>
+            intrinsic_permeability_models,
+        std::vector<std::unique_ptr<MaterialLib::PorousMedium::Porosity>>&&
+            porosity_models,
+        std::vector<std::unique_ptr<MaterialLib::PorousMedium::Storage>>&&
+            storage_models,
+        std::vector<std::unique_ptr<
+            MaterialLib::PorousMedium::CapillaryPressureSaturation>>&&
+            capillary_pressure_models,
+        std::vector<
+            std::unique_ptr<MaterialLib::PorousMedium::RelativePermeability>>&&
+            relative_permeability_models);
+
+    void setMaterialID(const ProcessLib::SpatialPosition& pos);
+
+    Eigen::MatrixXd const& getPermeability(
+        const double t,
+        const ProcessLib::SpatialPosition& pos,
+        const int dim) const;
+
+    double getPorosity(const double t, const ProcessLib::SpatialPosition& pos,
+                       const double p, const double T,
+                       const double porosity_variable) const;
+
+    double getNonwetRelativePermeability(const double t,
+                                         const ProcessLib::SpatialPosition& pos,
+                                         const double p, const double T,
+                                         const double saturation) const;
+    double getWetRelativePermeability(const double t,
+                                      const ProcessLib::SpatialPosition& pos,
+                                      const double p, const double T,
+                                      const double saturation) const;
+    double getCapillaryPressure(const double t,
+                                const ProcessLib::SpatialPosition& pos,
+                                const double p, const double T,
+                                const double saturation) const;
+    double getDerivCapillaryPressure(const double t,
+                                     const ProcessLib::SpatialPosition& pos,
+                                     const double p, const double T,
+                                     const double saturation) const;
+    double getLiquidDensity(const double p, const double T) const;
+    double getGasDensity(const double p, const double T) const;
+    double getGasViscosity(const double p, const double T) const;
+    double getLiquidViscosity(const double p, const double T) const;
+    double getDerivGasDensity(double const p, double const T) const;
+    bool computeConstitutiveRelation(
+        double const t,
+        ProcessLib::SpatialPosition const& x_position,
+        double const PG,
+        double const X,
+        double& Sw,
+        double& X_m,
+        double& dsw_dpg,
+        double& dsw_dX,
+        double& dxm_dpg,
+        double& dxm_dX);
+
+protected:
+    /// A flag to indicate whether the reference member, _material_ids,
+    /// is not assigned.
+    const bool _has_material_ids;
```

Comment:
I think it would be better to wrap _material_ids in boost::optional in this case.